src/controller/ArchipelagoController.ts

```diff
@@ -0,0 +1,212 @@
+
+  activeWorkerRequests: Record<string, PendingWorkerRequest> = {}
+
+  messageHandler: (m: WorkerMessage) => boolean
+
+  requestIdGenerator: IdGenerator = sequentialIdGenerator("")
+
+  options: WorkerControllerOptions
+
+  constructor(
+    messageHandler: (m: WorkerMessage) => any,
+    parameters: ArchipelagoParameters,
+    options: Partial<WorkerControllerOptions> = {}
+  ) {
+    this.worker = fork("./dist/worker/worker.js", [
+      JSON.stringify({ archipelagoParameters: parameters, logging: options.workerLogging ?? true }),
+    ])
+
+    this.worker.on("message", this.handleWorkerMessage.bind(this))
+
+    this.messageHandler = messageHandler
+
+    this.options = { requestTimeoutMs: 10 * 1000, ...options }
+  }
+
+  handleWorkerMessage(message: WorkerMessage) {
+    this.messageHandler(message)
+
+    if (message.type === "worker-status") {
+      this.workerStatus = message.status
+    } else if (message.type === "worker-request-error") {
+      const { requestId, error } = message
+      this.activeWorkerRequests[requestId]?.reject(error)
+      delete this.activeWorkerRequests[requestId]
+    } else if ("requestId" in message) {
+      const { requestId, payload } = message as WorkerResponse
+      this.activeWorkerRequests[requestId]?.resolve(payload)
+      delete this.activeWorkerRequests[requestId]
+    }
+  }
+
+  sendMessageToWorker(message: WorkerMessage) {
+    this.worker.send(message)
+  }
+
+  sendRequestToWorker<T>(message: Omit<WorkerRequest, "requestId">) {
+    const requestId = this.requestIdGenerator.generateId()
+
+    return new Promise<T>((resolve, reject) => {
+      this.activeWorkerRequests[requestId] = { resolve, reject }
+
+      this.sendMessageToWorker({ ...message, requestId } as WorkerMessage)
+
+      setTimeout(() => {
+        if (this.activeWorkerRequests[requestId]) {
+          delete this.activeWorkerRequests[requestId]
+          reject(new Error("Request timed out"))
+        }
+      }, this.options.requestTimeoutMs)
+    })
+  }
+
+  async dispose() {
+    await this.sendRequestToWorker({ type: "dispose-request" })
+    this.worker.kill()
+  }
+}
+
+export class ArchipelagoControllerImpl implements ArchipelagoController {
+  pendingUpdates: Map<string, PeerUpdate> = new Map()
+
+  updatesSubscribers: Set<UpdateSubscriber> = new Set()
+
+  activePeers: Map<string, PeerData> = new Map()
+  flushFrequency: number
+  logger: Logger
+
+  workerController: WorkerController
+
+  disposed: boolean = false
+
+  constructor(options: ArchipelagoControllerOptions) {
+    this.flushFrequency = options.flushFrequency ?? 2
+    this.logger = options.logger ?? console
+    this.workerController = new WorkerController(this.handleWorkerMessage.bind(this), options.archipelagoParameters)
+
+    this.startFlushLoop()
+  }
+
+  startFlushLoop() {
+    const loop = () => {
+      if (!this.disposed) {
+        this.flush()
+        setTimeout(loop, this.flushFrequency * 1000)
```

Comment:
Would it make sense to have a little more complex loop. I mean, let's assume that we use a 5 second interval, and that the processing took 5.1 seconds. I wouldn't want to wait 10 seconds for it to be sent again.

Does it make sense?

Again, can be done in another PR

Reply:
Makes sense. I'll be doing this in this PR. Shouldn't be too hard.